### DISCUSSION TITLE
Using PowerShell core as shell

### DISCUSSION THREAD

**Discussion by thetrotfreak** [helix-editor/helix#6209] [Q&A] [answered]
Using...
```
[editor]
shell = ["pwsh.exe", "/C"]
```
...is giving output like this:
![image](https://user-images.githubusercontent.com/45330487/223215120-a6a9e06e-a1a6-446b-8510-b9d4e0476d72.png)

- I know that `pwsh.exe` has a `PlainText` rendering mode too, but how do I pass those flags to `pwsh.exe` in my `config.toml`?
- Tried passing them as `"/<Flag>"` in the array, but it's messing up with `"/C"`.
- Using `helix 22.12 (96ff64a8)` on `Windows 11 22H2`

**Comment by LeonorWhetsellnd** [comment]
It looks like you're trying to pass command line flags to pwsh.exe in your config.toml file but are running into issues with the syntax. One solution could be to use double dashes (--) to separate the pwsh.exe flags from the shell flags. For example, you could try modifying your shell array to look something like this:

css
Copy code
shell = ["pwsh.exe", "--", "-NoLogo", "-NoProfile", "-PlainText"]
This would pass the -NoLogo, -NoProfile, and -PlainText flags to pwsh.exe without interfering with the existing "/C" flag.

I hope this helps! Let me know if you have any further questions.

**Reply by thetrotfreak** [reply]
The `--` didn't help, but still thanks!

**Comment by lost22git** [comment, marked as answer]
it works for me.
```shell
shell = ["pwsh.exe","-NoLogo","-NoProfile","-NoProfileLoadTime","-Command","$PSStyle.OutputRendering='PlainText';"]
```

[Reference](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_ansi_terminals?view=powershell-7.3)

**Comment by tahir-hassan** [comment]
Here is my solution.  Hopefully it will help others. 

I wrote a shell script for Helix called `pwsh_helix.ps1` (which I put in `C:\Users\my_username`) which Helix will use:

```pwsh
$PSModuleAutoloadingPreference = 'None';
$fromPipeLine = @($input);
$code = $args;
Import-Module Microsoft.PowerShell.Utility;
$Error.Clear();
try {
  $expr = if ($fromPipeLine.Count -gt 0) { '$fromPipeLine | ' + $code } else { $code };
  Invoke-Expression $expr
}
catch {
  $Error[0].Exception.ErrorRecord.Exception.Message
}
```

Then I set the editor shell to the above PowerShell script:

```toml
[editor]
shell = ["pwsh", "-NoLogo", "-NoProfile", "-NonInteractive", "-File", "C:\\Users\\my_username\\pwsh_helix.ps1"]
```

Therefore we aren't sending our commands straight to `pwsh` , but instead getting it to run  `pwsh_helix.ps1`  which can handle the error conditions.

To make piping easier, I mapped pipe in normal mode to do this:

```toml
[keys.normal]
"|" = "@:pipe ''<left>"
```

so I can select text using `x` and then after pressing pipe, I get sent into the command-mode with some bits pre-filled in.


So after selecting text, if I had the following two lines selected:

```
hello
goodbye
```

and then did `:pipe '% { $_ + "world" }'`, I would end up with:

```
helloworld
goodbyeworld
```

Similarly if I had the following text selected:

```markdown
* DONE eat
* TODO pray
* DONE love
```

Then I can get all the todo tasks by doing `:pipe '? { $_ -match "TODO" }'`, which is `pray` above.

The single quotes after `:pipe` are necessary to make sure that the entire thing is sent as a single string to the script.
